src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers/BinaryFormatterUnsafeDeserializeAnalyzer.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class BinaryFormatterUnsafeDeserializeAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "UA0012";
+        private const string Category = "Upgrade";
+
+        private const string TargetTypeSymbolName = "System.Runtime.Serialization.Formatters.Binary.BinaryFormatter";
+        private const string TargetMember = "UnsafeDeserialize";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.BinaryFormatterUnsafeDeserializeTitle), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.BinaryFormatterUnsafeDeserializeMessageFormat), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.BinaryFormatterUnsafeDeserializeDescription), Resources.ResourceManager, typeof(Resources));
+
+        private static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterSyntaxNodeAction(AnalyzeMemberAccessExpressions, SyntaxKind.SimpleMemberAccessExpression);
```

Comment:
Unfortunately, I think you'll need to analyze invocation expressions instead of member access expressions. UnsafeDeserialize takes two arguments whereas Deserialize only takes one, so you're going to have to change the invocation's arguments.

In most cases when UnsafeDeserialize is used, the second argument is null and it's ok to omit it when calling Deserialize instead. So, the change here is probably to only produce diagnostics for calls to UnsafeDeserialize that have a NullLiteralExpression for the second argument. Then update the code fix provider to remove the second (null) argument.

Reply:
We should raise the diagnostic regardless, but only have the codefixer do anything if it is not null

Reply:
Agreed

Reply:
Updated the unit tests. As-is, the analyzer will surface the diagnostic. This recommendation will be applied to  the code fixer.